### PR TITLE
char to_digit: avoid unnecessary casts to u64

### PR DESCRIPTION
**Pull request by Marcondiro**
Hello,
in the `char::to_digit` method there are a few `as u64` casts that are not strictly necessary.
I assume that the reason behind these casts is to avoid possible overflows in the `+ 10` add.

This PR removes the aforementioned casts, avoiding the overflow issue by slightly modifying the ASCII letter to int mapping.

Thanks,
Happy new year.

### CODE REVIEW

**Comment by jhpratt** [comment]
A quick look shows that the assembly is different as expected, so I'm going to kick off a perf run to verify things. If anything it should be slightly faster, but I know better than to predict perf blindly.

@bors try @rust-timer queue

**Comment by klensy** [comment]
cc @programmerjake #132709

**Comment by jhpratt** [comment]
Regressions seem to be noise, with there being minor improvements in other (less reliable) metrics. Even if it were neutral that would be sufficient for me.

Thanks!

@bors r+

**Comment by programmerjake** [comment]
does perf actually run the [`to_digit`  benchmarks](https://github.com/rust-lang/rust/blob/d117b7f211835282b3b177dc64245fff0327c04c/library/core/benches/char/methods.rs#L6-L37)? i didn't see them in the perf results...

**Comment by jhpratt** [comment]
@bors r-

Let's discuss on Zulip.

**Comment by klensy** [comment]
> does perf actually run the [`to_digit` benchmarks](https://github.com/rust-lang/rust/blob/d117b7f211835282b3b177dc64245fff0327c04c/library/core/benches/char/methods.rs#L6-L37)? i didn't see them in the perf results...

perf runs only on crates listed on page, smaller benchmarks from library not benched that way.

**Comment by programmerjake** [comment]
TLDR: this PR makes it measurably slower for the most common cases (fixed radix) on my computer

I manually ran the `to_digit` benchmarks on my AMD Ryzen 7950X:
```bash
for b in master pull-34969; do git switch "$b"; for i in {1..3}; do RUST_BACKTRACE=1 ./x.py bench library/core -- to_digit > "$b".txt; done; done
diff -U3 master.txt pull-34969.txt
```
master is at d117b7f211835282b3b177dc64245fff0327c04c, pull-34969 is at aa685bcccaa8fdb187e68de29f35db11aa99630f

```diff
--- master.txt  2024-12-31 16:52:47.377514215 -0800
+++ pull-34969.txt      2024-12-31 16:54:43.787387531 -0800
@@ -11,15 +11,15 @@
 bbbbb

 benchmarks:
-    char::methods::bench_to_digit_radix_10   2856.63ns/iter  +/- 20.97
-    char::methods::bench_to_digit_radix_16   2845.83ns/iter  +/- 13.06
-    char::methods::bench_to_digit_radix_2    2843.90ns/iter   +/- 9.80
-    char::methods::bench_to_digit_radix_36   9652.53ns/iter  +/- 54.87
-    char::methods::bench_to_digit_radix_var 15246.25ns/iter +/- 283.04
+    char::methods::bench_to_digit_radix_10   2882.53ns/iter  +/- 9.13
+    char::methods::bench_to_digit_radix_16   2885.77ns/iter  +/- 5.28
+    char::methods::bench_to_digit_radix_2    2882.67ns/iter  +/- 3.06
+    char::methods::bench_to_digit_radix_36   9654.94ns/iter +/- 38.77
+    char::methods::bench_to_digit_radix_var 14723.90ns/iter +/- 41.44

-test result: ok. 0 passed; 0 failed; 0 ignored; 5 measured; 514 filtered out; finished in 1.24s
+test result: ok. 0 passed; 0 failed; 0 ignored; 5 measured; 514 filtered out; finished in 5.39s

-       finished in 1.255 seconds
+       finished in 5.413 seconds
```

**Comment by programmerjake** [comment]
ran it another time and the results reversed for radix 2, 10, and 16:
```diff
-    char::methods::bench_to_digit_radix_10   2897.59ns/iter +/- 10.50
-    char::methods::bench_to_digit_radix_16   2884.83ns/iter  +/- 6.94
-    char::methods::bench_to_digit_radix_2    2883.97ns/iter +/- 11.17
-    char::methods::bench_to_digit_radix_36   9758.82ns/iter  +/- 5.09
-    char::methods::bench_to_digit_radix_var 15222.17ns/iter +/- 27.43
+    char::methods::bench_to_digit_radix_10   2843.65ns/iter +/- 13.94
+    char::methods::bench_to_digit_radix_16   2853.59ns/iter +/- 17.59
+    char::methods::bench_to_digit_radix_2    2850.38ns/iter +/- 15.10
+    char::methods::bench_to_digit_radix_36   9544.87ns/iter +/- 58.91
+    char::methods::bench_to_digit_radix_var 14527.29ns/iter +/- 38.01
```

so looks like just measurement noise to me for those and a distinct improvement for the `var` case, so r+ from me.

**Comment by jhpratt** [comment]
Looks neutral for the common case and an improvement for the variable case, then. A 1%ish difference in either direction is quite small and difficult to detect.

@bors r=jhpratt,programmerjake
